lib/chat/sync/usb_drive_file_dumper.ex

```diff
@@ -0,0 +1,121 @@
+defmodule Chat.Sync.UsbDriveFileDumper do
+  @moduledoc """
+  Saves file as a message.
+  """
+
+  alias Chat.{ChunkedFiles, ChunkedFilesMultisecret}
+  alias Chat.Db.ChangeTracker
+  alias Chat.{FileIndex, Identity, Log, Messages, Rooms, TaskSupervisor}
+  alias Chat.Sync.UsbDriveDumpFile
+  alias Chat.Upload.UploadKey
+  alias Phoenix.PubSub
+
+  @chunk_size Application.compile_env(:chat, :file_chunk_size)
+
+  def dump(%UsbDriveDumpFile{} = file, room_key, %Identity{} = room_identity) do
+    timestamp =
+      file.datetime
+      |> DateTime.from_naive!("Etc/UTC")
+      |> DateTime.to_unix()
+
+    type =
+      file
+      |> Map.get(:name)
+      |> Path.extname()
+      |> String.slice(1..-1)
+      |> String.downcase()
+      |> MIME.type()
+
+    destination = %{
+      pub_key: Base.encode16(room_key, case: :lower),
+      type: :room
+    }
+
+    entry = %{
+      client_last_modified: timestamp,
+      client_name: file.name,
+      client_relative_path: file.path,
+      client_size: file.size,
+      client_type: type
+    }
+
+    file_key = UploadKey.new(destination, room_key, entry)
+
+    file_awaiter =
+      Task.Supervisor.async(TaskSupervisor, fn ->
+        ChangeTracker.await({:file, file_key})
+      end)
+
+    file_secret = ChunkedFiles.new_upload(file_key)
+    ChunkedFilesMultisecret.generate(file_key, file.size, file_secret)
+
+    last_chunk_key =
+      file.path
+      |> File.stream!([], @chunk_size)
+      |> Stream.with_index()
+      |> Enum.map(fn {chunk, index} ->
+        chunk_start = @chunk_size * index
+
+        chunk_end =
+          if(@chunk_size * (index + 1) < file.size - 1,
+            do: @chunk_size * (index + 1),
+            else: file.size - 1
+          )
+
+        size = chunk_end - chunk_start
```

Comment:
`save_upload_chunk` needs file size, not a `chunk_size - 1`

it should be `size = file.size` to work correctly

Reply:
Ah 🤦‍♂️
You're right, it needs full file size, not chunk size.